### DISCUSSION TITLE
Websocket endpoint not available to clients

### DISCUSSION THREAD

**Discussion by S0mbre** [fastapi/fastapi#6283] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
#------ main.py -------#
from fastapi import FastAPI, WebSocket, WebSocketDisconnect

app = FastAPI(title='title', root_path='/api')

from fastapi.middleware.cors import CORSMiddleware

app.add_middleware(
    CORSMiddleware,
    allow_origins=['*'],
    allow_credentials=True,
    allow_methods=['*'],
    allow_headers=['*'],
)

@app.websocket('/wstest')
async def ws_test(websocket: WebSocket):
    try:
        print(f'WS: CONNECTING TO {websocket.url}, HDR = {websocket.headers}, PARAMS = {websocket.query_params} ...')
        await websocket.accept()
        print(f'WS: CONNECTED {websocket.url}')

        async for txt in websocket.iter_text():
            await websocket.send_text(txt)
    except WebSocketDisconnect as err:
        print(f'Websocket [{err.code}]: {err.reason}')
        raise

#------ Docker command -------#
gunicorn --bind 0.0.0.0:8000 -k uvicorn.workers.UvicornWorker main:app
```

### Description

The FastAPI backend runs from a Docker container on `HOST:8090`. 

The connection URL would be `ws://HOST:8090/api/wstest`.
Trying to connect to the websocket from [this test client](https://websocketking.com/), I get error `404`:
```
178.57.200.254 - - [16/Aug/2022:22:53:05 +0000] "GET /api/wstest HTTP/1.1" 404 22 "-" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/102.0.5005.167 YaBrowser/22.7.3.822 Yowser/2.5 Safari/537.36"
```
The browser log says one single word as error: "update".

OK, maybe that's just because of the server prefix (`/api`). Let's do without the prefix: `ws://HOST:8090/api/wstest`. Now the server returns `200`, but connection is still not established:
```
178.57.200.254 - - [16/Aug/2022:22:53:19 +0000] "GET /wstest HTTP/1.1" 200 710 "-" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/102.0.5005.167 YaBrowser/22.7.3.822 Yowser/2.5 Safari/537.36"
```

### Operating System

Linux

### FastAPI Version

0.78.0

### Python Version

3.10.2

### Additional Context

The app runs on the server behind a Nginx proxy with the following configuration:
```
events {}

http {
    server {
        listen 80;
        root  /usr/share/nginx/html;
        include /etc/nginx/mime.types;
        client_max_body_size 20M;

        location / {
            try_files $uri /index.html;
        }

        location /api/ {
            proxy_pass          http://use_backend:8000/;
            proxy_http_version  1.1;
            proxy_set_header    Host             $host;
            proxy_set_header    X-Real-IP        $remote_addr;
            proxy_set_header    X-Forwarded-For  $proxy_add_x_forwarded_for;
            proxy_read_timeout 1800;
            proxy_connect_timeout 1800;
        }

    }
}
```

I've got some slight suspicion that this may be the problem: Nginx doesn't allow ws connections.

As an apart, I've done everything suggested in [this topic](https://github.com/tiangolo/fastapi/issues/98), namely:

- used `APIRouter()`, then added to `app` routers => this was the initial setup, and failed
- created an extra `FastAPI` instance, then did `app.mount("/ws", ws)` => didn't help
- finally, made this test endpoint directly from `app` (as shown in example code) => same issues

It appears that as such, clients don't even reach `await websocket.accept()`, so the endpoint is entirely unavailable.

PS. My browser is Yandex Browser 22.7.3.822 (64-bit) - latest Chrome-based.

**Comment by iudeen** [comment, marked as answer]
I don't see an upgrade statement in NGINX conf. I suspect that to be the culprit. Check this [blogpost](https://www.nginx.com/blog/websocket-nginx/)

Example:
```
location / {
  proxy_pass http://localhost:8080;
  proxy_http_version 1.1;
  proxy_set_header Upgrade $http_upgrade;
  proxy_set_header Connection "upgrade";
  proxy_set_header Host $host;
}
```

**Comment by S0mbre** [comment]
In fact, that WAS Nginx. This config has done my day:
```
events {}

http {
    map $http_upgrade $connection_upgrade {
        default upgrade;
        '' close;
    }

    server {
        listen 80;
        root  /usr/share/nginx/html;
        include /etc/nginx/mime.types;
        client_max_body_size 20M;

        location / {
            try_files $uri /index.html;
        }

        location /api/ {
            # WHATEVER...
        }

        location /ws/ {
            proxy_pass                      http://<HOST>:8000;
            proxy_http_version              1.1;
            proxy_set_header                Upgrade          $http_upgrade;
            proxy_set_header                Connection       "Upgrade";
            proxy_set_header                Host             $host;
            proxy_set_header                X-Real-IP        $remote_addr;
            proxy_set_header                X-Forwarded-For  $proxy_add_x_forwarded_for;
            proxy_read_timeout              1800;
            proxy_connect_timeout           1800;
        }

    }
}
```

Issue closed. Thanks @iudeen
